File: ui/src/api/anomaly.ts
import apiClient from './client'
import type { PaginatedResponse } from './types'

export interface AnomalyAlert {
  id: number
  host_id: string
  hostname: string
  alert_type: 'isolation_forest' | 'correlation'
  pattern_name: string
  severity: 'critical' | 'high' | 'medium' | 'low'
  anomaly_score: number
  top_metric: string
  top_value: number
  description: string
  status: 'open' | 'confirmed' | 'false_positive'
  resolved_by: string
  created_at: string
  updated_at: string
}

export interface AnomalyStats {
  total: number
  open: number
  critical: number
  by_type: { alert_type: string; count: number }[]
  by_pattern: { alert_type: string; count: number }[]
}

export interface ListAnomalyParams {
  page?: number
  page_size?: number
  host_id?: string
  alert_type?: string
  severity?: string
  status?: string
}

export const anomalyApi = {
  list: (params?: ListAnomalyParams) => {
    return apiClient.get<PaginatedResponse<AnomalyAlert>>('/anomalies', { params })
  },

  stats: () => {
    return apiClient.get<AnomalyStats>('/anomalies/stats')
  },

  resolve: (id: number, status: 'confirmed' | 'false_positive') => {
    return apiClient.put(`/anomalies/${id}/resolve`, { status })
  },
}

File: internal/server/model/anomaly.go
package model

// AnomalyAlert represents an ML-detected behavioral anomaly.
type AnomalyAlert struct {
	ID           uint      `gorm:"primarykey" json:"id"`
	HostID       string    `gorm:"type:varchar(64);index" json:"host_id"`
	Hostname     string    `gorm:"type:varchar(255)" json:"hostname"`
	AlertType    string    `gorm:"type:varchar(50);index" json:"alert_type"`        // isolation_forest / correlation
	PatternName  string    `gorm:"type:varchar(100)" json:"pattern_name"`           // correlation pattern name (c2_beacon, etc)
	Severity     string    `gorm:"type:varchar(20);default:medium" json:"severity"` // critical/high/medium/low
	AnomalyScore float64   `gorm:"type:double;default:0" json:"anomaly_score"`      // 0.0-1.0
	TopMetric    string    `gorm:"type:varchar(100)" json:"top_metric"`             // most anomalous metric name
	TopValue     float64   `gorm:"type:double;default:0" json:"top_value"`          // observed value of top metric
	Description  string    `gorm:"type:text" json:"description"`
	Status       string    `gorm:"type:varchar(20);default:open" json:"status"` // open/confirmed/false_positive
	ResolvedBy   string    `gorm:"type:varchar(100)" json:"resolved_by,omitempty"`
	CreatedAt    LocalTime `json:"created_at"`
	UpdatedAt    LocalTime `json:"updated_at"`
}

func (AnomalyAlert) TableName() string { return "anomaly_alerts" }
